### PR TITLE
[Feature](Nereids)add rule to change scalar sub-query to window function

### PR DESCRIPTION
**Pull request by sohardforaname**
# Proposed changes

Issue Number: close #xxx

## Problem summary

SQL like TPC-H Q2 and Q17, which contains a correlated sub-query with only one aggregation function output, we can eliminate the sub-query and transform it to window function. For example, TPC-H Q17 is

```sql
select
        sum(l_extendedprice) / 7.0 as avg_yearly
    from
        lineitem,
        part
    where
        p_partkey = l_partkey
        and p_brand = 'Brand#23'
        and p_container = 'MED BOX'
        and l_quantity < (
            select
                0.2 * avg(l_quantity)
            from
                lineitem
            where
                l_partkey = p_partkey
        );
```

we change it to

```sql
select
        sum(l_extendedprice) / 7.0 as avg_yearly
    from (
    select
        l_extendedprice, l_quantity, avg(l_quantity) over(partition by l_partkey) avg_l_quantity
    from 
        lineitem,
        part
    where
        p_partkey = l_partkey
        and p_brand = 'Brand#23'
        and p_container = 'MED BOX' )
    where l_quantity < 0.2 * avg_l_quantity
```

now the rule can only handle: where conjuncts in outer scope contain one sub-query and the conjunct contain sub-query is a comparison-predicate, we will support compound-predicate and more than one conjuncts containing sub-query later.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [x] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [x] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run fe ut

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run p0

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+        if (project.child() == null || !(project.child() instanceof LogicalApply)) {
+            return false;
+        }
+        // filter(apply()) may be also ok.
+        // but nereids will match the pattern filter(project(apply()))
+        LogicalApply apply = ((LogicalApply<?, ?>) project.child());
+        if (!apply.isScalar() || !apply.isCorrelated() || !apply.getSubCorrespondingConject().isPresent()) {
+            return false;
+        }
+        return apply.left() != null && apply.right() instanceof LogicalAggregate
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java:139]: it is not nessesary to check apply.left() != null

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java:1]: fix warning hints by idea as mush as possible.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+                .collect(Collectors.toList());
+        List<LogicalRelation> innerTables = innerPlans.stream().filter(LogicalRelation.class::isInstance)
+                .map(LogicalRelation.class::cast)
+                .collect(Collectors.toList());
+
+        Set<Long> outerIds = outerTables.stream().map(node -> node.getTable().getId()).collect(Collectors.toSet());
+        Set<Long> innerIds = innerTables.stream().map(node -> node.getTable().getId()).collect(Collectors.toSet());
+
+        if (outerTables.size() != outerIds.size() || innerTables.size() != innerIds.size()
+                || outerIds.size() != innerIds.size() + 1) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java:240]: why we need them ? `outerTables.size() != outerIds.size() || innerTables.size() != innerIds.size()`

Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java:240]: why `outerIds.size() != innerIds.size() + 1`? paper do not has this restirct

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+            return false;
+        }
+
+        outerIds.removeAll(innerIds);
+        if (outerIds.size() != 1) {
+            return false;
+        }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java:247]: only need to check outer table contains all inner table, and outer table that do not appear in inner have no join condition with the table correlated with inner.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+        aggOp = aggSet.get(0);
+        functions = ((List<AggregateFunction>) ExpressionUtils.<AggregateFunction>collectAll(
+                aggOp.getOutputExpressions(), AggregateFunction.class::isInstance));
+        Preconditions.checkArgument(functions.size() == 1);
+        return functions.stream().allMatch(f -> SUPPORTED_FUNCTION.contains(f.getClass()) && !f.isDistinct());
+    }
+
+    // check if the relations of the outer's includes the inner's
+    private boolean checkRelation(List<Expression> correlatedSlots) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java:228]: current code in check relation is too tight. we could loose it in future

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+
+    private boolean checkPredicate(Set<Expression> outerConjuncts, Set<Expression> innerConjuncts) {
+        Iterator<Expression> innerIter = innerConjuncts.iterator();
+        // inner predicate should be the sub-set of outer predicate.
+        while (innerIter.hasNext()) {
+            Expression innerExpr = innerIter.next();
+            Iterator<Expression> outerIter = outerConjuncts.iterator();
+            while (outerIter.hasNext()) {
+                Expression outerExpr = outerIter.next();
+                if (new ExpressionIdenticalChecker().check(innerExpr, outerExpr)) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java:264]: use static instance instead new

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+        LogicalAggregate<Plan> agg = apply.right();
+
+        // for example, in tpc-h Q17
+        // window filter conjuncts is cast(l_quantity#id1 as decimal(27, 9)) < `0.2 * avg(l_quantity)`#id2 and
+        // 0.2 * avg(l_quantity#id3) as `0.2 * l_quantity`#id2 of agg's output expression
+        // we change it to cast(l_quantity#id1 as decimal(27, 9)) < 0.2 * `avg(l_quantity#id1) over(window)`#id4
+        // and avg(l_quantity#id1) over(window) as `avg(l_quantity#id1) over(window)`#id4
+        // first we get the slots of the two sides of apply
+
+        Preconditions.checkArgument(apply.getSubCorrespondingConject().get() instanceof ComparisonPredicate);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java:172]: getSubCorrespondingConject fix this typo

Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java:172]: why check here? where to ensure `apply.getSubCorrespondingConject()` is not empty and it  must be a `ComparisonPredicate`?
if no where to ensure this.
we should add this to check and return false if it is not satisfy the restriction

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+
+        // adjust agg function's nullable.
+        AggregateFunction function = functions.get(0);
+        if (function instanceof NullableAggregateFunction) {
+            function = ((NullableAggregateFunction) function).withAlwaysNullable(false);
+        }
+
+        WindowExpression windowFunction = createWindowFunction(apply.getCorrelationSlot(),
+                function.withChildren(windowAggSlots));
+        NamedExpression windowFunctionAlias = new Alias(windowFunction, "wf");
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java:188]: u need use a unique name generate by column name generator
refer to `statementContext.generateColumnName()`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+        // and avg(l_quantity#id1) over(window) as `avg(l_quantity#id1) over(window)`#id4
+        // first we get the slots of the two sides of apply
+
+        Preconditions.checkArgument(apply.getSubCorrespondingConject().get() instanceof ComparisonPredicate);
+        ComparisonPredicate windowFilterConjunct = PlanUtils.maybeCommuteComparisonPredicate(
+                ((ComparisonPredicate) apply.getSubCorrespondingConject().get()), apply.left());
+
+        // build window function, replace the slot
+        List<Expression> windowAggSlots = windowFilterConjunct.left().collectToList(Slot.class::isInstance);
+        Preconditions.checkArgument(windowAggSlots.size() == 1);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java:178]: we should not check, unless u have verify it before or `windowAggSlots.size() != 1` represent a bug. if so, u need add some error msg

**Review by morrySnow — CHANGES_REQUESTED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+        LogicalPlan outer = ((LogicalPlan) apply.child(0));
+        LogicalPlan inner = ((LogicalPlan) apply.child(1));
+        outerPlans = new PlanCollector().collect(outer);
+        innerPlans = new PlanCollector().collect(inner);
+        LogicalFilter outerFilter = node;
+        Optional<LogicalFilter> innerFilter = innerPlans.stream()
+                .filter(LogicalFilter.class::isInstance)
+                .map(LogicalFilter.class::cast).findFirst();
+        return innerFilter.filter(
+                logicalFilter -> checkPlanType() && checkAggType()
```
[the remark this answers is not in the document]
Comment (sohardforaname) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java:154]: idea recommends it

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java [hunk 10]
```diff
@@ -0,0 +1,387 @@
...
+        if (!check(filter, apply)) {
+            return filter;
+        }
+        return trans(filter, apply);
+    }
+
+    private LogicalApply<Plan, LogicalAggregate<Plan>> checkPattern(LogicalFilter<? extends Plan> filter) {
+        if (!(filter.child() instanceof LogicalProject)) {
+            return null;
+        }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java:115]: should we need to remove these lines?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java [hunk 11]
```diff
@@ -0,0 +1,387 @@
...
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * change the plan:
+ * logicalFilter(logicalApply(any(), logicalAggregate()))
+ * to
+ * logicalProject((logicalFilter(logicalWindow(logicalFilter(any())))))
+ */
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java:78]: add refer paper here

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run  buildall

**Review by morrySnow — APPROVED** [review]

**Pull request merged**